tools/hsmtool.c

```diff
@@ -368,6 +371,55 @@ static int guess_to_remote(const char *address, struct node_id *node_id,
 	return 1;
 }
 
+static int dumponchaindescriptors(const char *hsm_secret_path, const char *passwd)
+{
+	struct secret hsm_secret;
+	u8 bip32_seed[BIP32_ENTROPY_LEN_256];
+	u32 salt = 0;
+	struct ext_key master_extkey;
+	char *enc_xpub, *descriptor, checksum[DESCRIPTOR_CHECKSUM_LENGTH + 1];
+
+	if (passwd)
+		get_encrypted_hsm_secret(&hsm_secret, hsm_secret_path, passwd);
+	else
+		get_hsm_secret(&hsm_secret, hsm_secret_path);
+
+	/* We use m/0/0/k as the derivation tree for onchain funds. */
+
+	/* The root seed is derived from hsm_secret using hkdf.. */
+	do {
+		hkdf_sha256(bip32_seed, sizeof(bip32_seed),
+			    &salt, sizeof(salt),
+			    &hsm_secret, sizeof(hsm_secret),
+			    "bip32 seed", strlen("bip32 seed"));
+		salt++;
+		/* ..Which is used to derive m/ */
+	} while (bip32_key_from_seed(bip32_seed, sizeof(bip32_seed),
+				     /* An xpub can easily be converted to a tpub */
+				     BIP32_VER_MAIN_PRIVATE,
+				     0, &master_extkey) != WALLY_OK);
+
+	if (bip32_key_to_base58(&master_extkey, BIP32_FLAG_KEY_PUBLIC, &enc_xpub) != WALLY_OK)
+		errx(ERROR_LIBWALLY, "Can't encode xpub");
+
+	/* Now we format the descriptor strings (we only ever create P2WPKH and
+	 * P2SH-P2WPKH outputs). */
```

Comment:
iirc, anchor outputs break this assumption

Reply:
We could end up with anything if we use the `close_to` arguments, that includes addresses that we don't even control ourselves. The tool makes it clear we only show on-chain addresses that were generated by the internal wallet, so I don't think this'll be an issue.

Reply:
Yeah, now we don't promote it as a recovery mechanism anyways :-)